WWST-955 Orvibo Motion Sensor

jinkangzhangS opened the pull request:
Add Orvibo Motion Sensor DTH .

jinkangzhangS committed:
add Orvibo Motion sensor DTH

add Orvibo Motion sensor DTH referring to Orvibo ZCL info.

jinkangzhangS committed:
add Orvibo Motion sensor DTH

add Orvibo Motion sensor DTH referring to Orvibo ZCL info.

greens commented on devicetypes/Orvibo/Orvibo-Motion-Sensor.src/Orvibo-Motion-Sensor.groovy:
```diff
@@ -0,0 +1,126 @@
+/**
+ *  Orvibo Motion Sensor
+ *
+ *  Copyright 2016 Chris Kitch
```
update this

greens commented on devicetypes/Orvibo/Orvibo-Motion-Sensor.src/Orvibo-Motion-Sensor.groovy:
```diff
@@ -0,0 +1,126 @@
+ *
+ */
+metadata {
+    definition (name: "Orvibo Motion Sensor", namespace: "Orvibo", author: "jk0218.zhang@samsung.com", vid: "generic-motion-2", ocfDeviceType: "x.com.st.d.sensor.motion") {
```
If the DTH was written in-house, the `namespace` and `author` should be `"smartthings"` and `"SmartThings"`.

Also the file should be moved to the `deviceTypes/smartthings` directory.

greens commented on devicetypes/Orvibo/Orvibo-Motion-Sensor.src/Orvibo-Motion-Sensor.groovy:
```diff
@@ -0,0 +1,126 @@
+metadata {
+    definition (name: "Orvibo Motion Sensor", namespace: "Orvibo", author: "jk0218.zhang@samsung.com", vid: "generic-motion-2", ocfDeviceType: "x.com.st.d.sensor.motion") {
+		capability "Motion Sensor"
+		capability "Battery"
```
you will need to add the "Health Check" capability (and set the `checkInterval` and provide a `ping()` method) in order to pass certification

greens commented on devicetypes/Orvibo/Orvibo-Motion-Sensor.src/Orvibo-Motion-Sensor.groovy:
```diff
@@ -0,0 +1,126 @@
+
+        fingerprint profileId: "0104", deviceId: "0402", inClusters: "0000,0003,0500,0001", outClusters: "", model: "895a2d80097f4ae2b2d40500d5e03dcc", manufacturer: "ORVIBO"
+	}
+    
```
You will need to add a basic tile definition in order for the device to function in ST Classic which is necessary for certification

greens commented on devicetypes/Orvibo/Orvibo-Motion-Sensor.src/Orvibo-Motion-Sensor.groovy:
```diff
@@ -0,0 +1,126 @@
+    
+    listResult << result
+    
+    if (listMap) {
```
never true

greens commented on devicetypes/Orvibo/Orvibo-Motion-Sensor.src/Orvibo-Motion-Sensor.groovy:
```diff
@@ -0,0 +1,126 @@
+	def descriptionText = null
+    log.debug "Parsing: ${description}"
+    
+	Map map = [:]
```
just use this map instead of having both a map and a separate name, value and descriptiontext

greens commented on devicetypes/Orvibo/Orvibo-Motion-Sensor.src/Orvibo-Motion-Sensor.groovy:
```diff
@@ -0,0 +1,126 @@
+	if (value == "active") {
+    	def timeout = 2
+        
+        if (motionStopTime)
```
You will need to put in extra code here to use the default value. The "default" in the preference definition does not actually set the value in memory.

greens commented on devicetypes/Orvibo/Orvibo-Motion-Sensor.src/Orvibo-Motion-Sensor.groovy:
```diff
@@ -0,0 +1,126 @@
+	Map map = [:]
+
+	List listMap = []
+	List listResult = []
```
the code below will never result in two events being in the result list

greens commented on devicetypes/Orvibo/Orvibo-Motion-Sensor.src/Orvibo-Motion-Sensor.groovy:
```diff
@@ -0,0 +1,126 @@
+}
+
+private Map parseReportAttributeMessage(String description) {
+	Map descMap = (description - "read attr - ").split(",").inject([:]) {
```
`zigbee.parseDescriptionAsMap()`

greens commented on devicetypes/Orvibo/Orvibo-Motion-Sensor.src/Orvibo-Motion-Sensor.groovy:
```diff
@@ -0,0 +1,126 @@
+	return resultMap
+}
+
+private Integer convertHexToInt(hex) {
```
if the method is only called once it probably doesn't need to be broken out.

greens reviewed (changes requested):
Aside from other changes: 
* please use consistent spacing (tabs or spaces)
* add a poll of the initial state in the `installed()` method

jinkangzhangS committed:
update Orvibo motion sensor.

update Orvibo motion sensor.

jinkangzhangS committed:
remove used code

update Orvibo motion sensor , removed used dth code

MAblewicz commented on devicetypes/smartthings/orvibo-motion-sensor.src/Orvibo-Motion-Sensor.groovy:
```diff
@@ -0,0 +1,200 @@
+		capability "Health Check"
+		capability "Sensor"
+
+		command "enrollResponse"
```
No neccessary and used. Remove it. Command doesn't work on SC

MAblewicz commented on devicetypes/smartthings/orvibo-motion-sensor.src/Orvibo-Motion-Sensor.groovy:
```diff
@@ -0,0 +1,200 @@
+
+		command "enrollResponse"
+
+		fingerprint inClusters: "0000,0001,0003,0500", outClusters: ""
```
No outClusters device so remove it. Add deviceJoinName

MAblewicz commented on devicetypes/smartthings/orvibo-motion-sensor.src/Orvibo-Motion-Sensor.groovy:
```diff
@@ -0,0 +1,200 @@
+import physicalgraph.zigbee.zcl.DataType
+
+metadata {
+	definition(name: "Orvibo Motion Sensor", namespace: "smartthings", author: "SmartThings", runLocally: false, minHubCoreVersion: '000.017.0012', executeCommandsLocally: false, mnmn: "SmartThings", vid: "generic-motion") {
```
vid:generic-motion is for devices with temperature measurement. Use VID: generic-motion-2 instead

MAblewicz commented on devicetypes/smartthings/orvibo-motion-sensor.src/Orvibo-Motion-Sensor.groovy:
```diff
@@ -0,0 +1,200 @@
+	}
+}
+
+private List<Map> collectAttributes(Map descMap) {
```
dont define return types and  access  modifiers . Its groovy just add "def"

MAblewicz commented on devicetypes/smartthings/orvibo-motion-sensor.src/Orvibo-Motion-Sensor.groovy:
```diff
@@ -0,0 +1,200 @@
+	Map map = zigbee.getEvent(description)
+	if (!map) {
+		if (description?.startsWith('zone status')) {
+			map = parseIasMessage(description)
```
Extract motion support to method, please don't create this big fat parse method if not necessary

MAblewicz commented on devicetypes/smartthings/orvibo-motion-sensor.src/Orvibo-Motion-Sensor.groovy:
```diff
@@ -0,0 +1,200 @@
+
+			}
+		} else {
+			Map descMap = zigbee.parseDescriptionAsMap(description)
```
Extract battery raporting to method

MAblewicz commented on devicetypes/smartthings/orvibo-motion-sensor.src/Orvibo-Motion-Sensor.groovy:
```diff
@@ -0,0 +1,200 @@
+	sendEvent(getMotionResult('inactive'))
+}
+
+private Map parseIasMessage(String description) {
```
"def"?

MAblewicz commented on devicetypes/smartthings/orvibo-motion-sensor.src/Orvibo-Motion-Sensor.groovy:
```diff
@@ -0,0 +1,200 @@
+	translateZoneStatus(zs)
+}
+
+private Map translateZoneStatus(ZoneStatus zs) {
```
"def"?

MAblewicz commented on devicetypes/smartthings/orvibo-motion-sensor.src/Orvibo-Motion-Sensor.groovy:
```diff
@@ -0,0 +1,200 @@
+	return (zs.isAlarm1Set() || zs.isAlarm2Set()) ? getMotionResult('active') : getMotionResult('inactive')
+}
+
+private Map getBatteryPercentageResult(rawValue) {
```
"def"?

MAblewicz commented on devicetypes/Orvibo/Orvibo-Motion-Sensor.src/Orvibo-Motion-Sensor.groovy:
```diff
@@ -0,0 +1,126 @@
+		}
+	}
+}
+
```
No initial state checking in installed() method. You need to do it to synchronize device state with platform

MAblewicz commented on devicetypes/smartthings/orvibo-motion-sensor.src/Orvibo-Motion-Sensor.groovy:
```diff
@@ -0,0 +1,200 @@
+
+	return  descMaps
+}
+
```
No initial state checking in installed() method. You need to do it to synchronize device state with platform

jinkangzhangS committed:
update orvibo dth

update orvibo dth

jinkangzhangS committed:
add model id for orvibo sensor.

add model id in fingerprint , fixed battery return null bug .

greens commented on devicetypes/smartthings/orvibo-motion-sensor.src/Orvibo-Motion-Sensor.groovy:
```diff
@@ -0,0 +1,207 @@
+	log.debug "configure"
+	// Device-Watch allows 2 check-in misses from device + ping (plus 1 min lag time)
+	// enrolls with default periodic reporting until newer 5 min interval is confirmed
+	sendEvent(name: "checkInterval", value:20 * 60, displayed: false, data: [protocol: "zigbee", hubHardwareId: device.hub.hardwareID, offlinePingable: "1"])
```
The comment above does not apply to this device.

Please add another minute to this value.

greens commented on devicetypes/smartthings/orvibo-motion-sensor.src/Orvibo-Motion-Sensor.groovy:
```diff
@@ -0,0 +1,207 @@
+	if (descMap?.clusterInt == zigbee.POWER_CONFIGURATION_CLUSTER && descMap.commandInt != 0x07 && descMap.value) {
+		log.info "BATT METRICS - attr: ${descMap?.attrInt}, value: ${descMap?.value}, decValue: ${Integer.parseInt(descMap.value, 16)}, currPercent: ${device.currentState("battery")?.value}, device: ${device.getDataValue("manufacturer")} ${device.getDataValue("model")}"
+		List<Map> descMaps = collectAttributes(descMap)
+	    def battMap = descMaps.find { it.attrInt == 0x0021 }
```
please use only tabs instead of a mix of tabs and spaces

greens commented on devicetypes/smartthings/orvibo-motion-sensor.src/Orvibo-Motion-Sensor.groovy:
```diff
@@ -0,0 +1,207 @@
+
+	simulator {
+		status "active": "zone status 0x0001 -- extended status 0x00"
+        for (int i = 0; i <= 100; i += 11) {
```
please fix whitespace

jinkangzhangS committed:
add ovribo motion sensor install inital state poll

[body]

jinkangzhangS committed:
using tab installed space .

jinkangzhangS committed:
remove fake attribute in install function

remove command configrure report .
the orvibo vendor did not support it .
The battery attribute is reported without bind and reporting CFG.
the NXP plan reports the electricity once in 20 minutes.

MAblewicz commented on devicetypes/smartthings/orvibo-motion-sensor.src/Orvibo-Motion-Sensor.groovy:
```diff
@@ -0,0 +1,167 @@
+}
+def batteyHandler(String description){
+	def descMap = zigbee.parseDescriptionAsMap(description)
+    def map
```
formatting

MAblewicz commented on devicetypes/smartthings/orvibo-motion-sensor.src/Orvibo-Motion-Sensor.groovy:
```diff
@@ -0,0 +1,167 @@
+}
+def configure() {
+	log.debug "configure"
+	sendEvent(name: "checkInterval", value:20 * 60, displayed: false, data: [protocol: "zigbee", hubHardwareId: device.hub.hardwareID, offlinePingable: "1"])
```
Due to possibility of network lag add some additional timeout to this like 2 mins

MAblewicz commented on devicetypes/smartthings/orvibo-motion-sensor.src/Orvibo-Motion-Sensor.groovy:
```diff
@@ -0,0 +1,167 @@
+def getMotionResult(value) {
+	String descriptionText = value == 'active' ? "${device.displayName} detected motion" : "${device.displayName} motion has stopped"
+	return [
+			name			: 'motion',
```
formatting

MAblewicz commented on devicetypes/smartthings/orvibo-motion-sensor.src/Orvibo-Motion-Sensor.groovy:
```diff
@@ -0,0 +1,167 @@
+	def descMap = zigbee.parseDescriptionAsMap(description)
+    def map
+	if (descMap?.clusterInt == zigbee.POWER_CONFIGURATION_CLUSTER && descMap.commandInt != 0x07 && descMap.value) {
+		log.info "BATT METRICS - attr: ${descMap?.attrInt}, value: ${descMap?.value}, decValue: ${Integer.parseInt(descMap.value, 16)}, currPercent: ${device.currentState("battery")?.value}, device: ${device.getDataValue("manufacturer")} ${device.getDataValue("model")}"
```
log.info goes to Sumologic. Please avoid it if not necessary.

MAblewicz commented on devicetypes/smartthings/orvibo-motion-sensor.src/Orvibo-Motion-Sensor.groovy:
```diff
@@ -0,0 +1,167 @@
+    def map
+	if (descMap?.clusterInt == zigbee.POWER_CONFIGURATION_CLUSTER && descMap.commandInt != 0x07 && descMap.value) {
+		log.info "BATT METRICS - attr: ${descMap?.attrInt}, value: ${descMap?.value}, decValue: ${Integer.parseInt(descMap.value, 16)}, currPercent: ${device.currentState("battery")?.value}, device: ${device.getDataValue("manufacturer")} ${device.getDataValue("model")}"
+		List<Map> descMaps = collectAttributes(descMap)
```
Refactor this part of code.  Check several things:
1. What is reported in descMap.attrInt  (if it is 0x0021 some of code is not necessary)
2. Is "collectAttributes()" required in that case. Does device reports anything in descMap.additionalAttrs?

MAblewicz commented on devicetypes/smartthings/orvibo-motion-sensor.src/Orvibo-Motion-Sensor.groovy:
```diff
@@ -0,0 +1,167 @@
+def motionHandler(String description){
+	//inactive
+	def isActive = zigbee.translateStatusZoneType19(description)
+	def value = isActive ? "active" : "inactive"
```
Combine two lines. Value is used only in IF. Why not use isActive variable there instead.

MAblewicz commented on devicetypes/smartthings/orvibo-motion-sensor.src/Orvibo-Motion-Sensor.groovy:
```diff
@@ -0,0 +1,167 @@
+}
+def parseIasMessage(String description) {
+	ZoneStatus zs = zigbee.parseZoneStatus(description)
+	translateZoneStatus(zs)
```
You are creating new method (translateZoneStatus) just to call return without any additional logic. Just call return here.
Or even better. Change getMotionResult(value) method. Check if value is true or false and create events based on that. Remeber to change stopMotion().

MAblewicz commented on devicetypes/smartthings/orvibo-motion-sensor.src/Orvibo-Motion-Sensor.groovy:
```diff
@@ -0,0 +1,167 @@
+			status "battery ${i}%": "read attr - raw: 2E6D01000108210020C8, dni: 2E6D, endpoint: 01, cluster: 0001, size: 08, attrId: 0021, encoding: 20, value: ${i}"
+		}
+	}
+	preferences {
```
Are those preferences required. They are only supported by ST Classic?

MAblewicz reviewed (dismissed):
Please refer to my comments in code.

jinkangzhangS committed:
optimize the dth code

greens commented on devicetypes/smartthings/orvibo-motion-sensor.src/Orvibo-Motion-Sensor.groovy:
```diff
@@ -0,0 +1,140 @@
+}
+def batteyHandler(String description){
+	def descMap = zigbee.parseDescriptionAsMap(description)
+    def map = [:]
```
fix whitespace

jinkangzhangS committed:
remove whitespace
